File: modules/kscan-live-vto-native/ios/Core/LiveVtoGeometrySnapshotJson.swift
import Foundation

/// Deterministic JSON serialization of a `GeometrySnapshot`.
///
/// Hand-rolled rather than `JSONSerialization`/`JSONEncoder` on purpose --
/// field-for-field port of Android's `LiveVtoGeometrySnapshotJson.kt`, whose
/// own header explains why a hand-rolled serializer is required there
/// (`org.json` is stubbed in JVM unit tests). That specific reason does not
/// apply on iOS (`JSONSerialization` works identically in `swift test` and in
/// the real app), but the OUTPUT FORMAT is kept byte-for-byte compatible with
/// Android's on purpose: fixed key order, full-precision floats, sorted
/// control-point keys -- so the two platforms' diagnostic snapshots (and any
/// future cross-platform diff tooling) are directly comparable without a
/// normalization step.
public enum GeometrySnapshotJson {

  public static func encode(_ s: GeometrySnapshot, includeMesh: Bool = false) -> String {
    var b = "{"
    b += str("fixtureId", s.fixtureId) + ","
    b += str("bodyFrameId", s.bodyFrameId) + ","
    b += str("activeAssetId", s.activeAssetId) + ","
    b += str("assetVersion", s.assetVersion) + ","
    b += "\"failure\":" + (s.failure.map(quote) ?? "null") + ","
    b += "\"gatePassed\":\(s.gatePassed),"
    b += "\"gateFindings\":[" + s.gateFindings.map(quote).joined(separator: ",") + "],"
    b += num("scale", s.scale) + ","
    b += num("rotationRadians", s.rotationRadians) + ","
    b += "\"controlPoints\":{"
    // sorted for determinism -- dictionary iteration order must not leak into evidence
    b += s.controlPoints.sorted { $0.key < $1.key }
      .map { "\(quote($0.key)):[\(fmt($0.value.x)),\(fmt($0.value.y))]" }
      .joined(separator: ",")
    b += "},"
    b += "\"bounds\":{"
    b += num("minX", s.boundsMin.x) + ","
    b += num("minY", s.boundsMin.y) + ","
    b += num("maxX", s.boundsMax.x) + ","
    b += num("maxY", s.boundsMax.y) + "},"
    b += num("canvasWidth", s.canvasWidth) + ","
    b += num("canvasHeight", s.canvasHeight) + ","
    b += "\"textureWidth\":\(s.textureWidth),"
    b += "\"textureHeight\":\(s.textureHeight),"
    b += "\"meshWidth\":\(s.meshWidth),"
    b += "\"meshHeight\":\(s.meshHeight),"
    b += "\"meshVertexCount\":\((s.meshVertices?.count ?? 0) / 2),"
    b += "\"validationProblems\":[" + s.validate().map(quote).joined(separator: ",") + "]"
    if includeMesh, let meshVertices = s.meshVertices {
      b += ",\"meshVertices\":[" + meshVertices.map(fmt).joined(separator: ",") + "]"
    }
    b += "}"
    return b
  }

  private static func fmt(_ f: Float) -> String { f.isFinite ? formatFloat(f) : "\"\(f)\"" }

  /// Swift's default `Float` string conversion and Kotlin's `Float.toString()`
  /// use different (but both round-trippable) formatting rules; this matters
  /// only for byte-identical diffing across platforms, never for the numeric
  /// value a test actually compares. Emits the shortest round-trippable
  /// decimal, matching `Double`'s well-defined `description` semantics.
  private static func formatFloat(_ f: Float) -> String {
    if f == f.rounded() && abs(f) < 1e15 {
      return String(format: "%.1f", f)
    }
    return "\(f)"
  }

  private static func quote(_ v: String) -> String {
    "\"" + v.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: "\"", with: "\\\"") + "\""
  }
  private static func str(_ k: String, _ v: String) -> String { "\(quote(k)):\(quote(v))" }
  private static func num(_ k: String, _ v: Float) -> String { "\(quote(k)):\(fmt(v))" }
}

/// Short one-line form for logging -- never carries mesh vertices.
public func describeSnapshot(_ s: GeometrySnapshot) -> String {
  GeometrySnapshotJson.encode(s, includeMesh: false)
}

File: modules/kscan-live-vto-native/android/src/main/java/expo/modules/kscanlivevtonative/LiveVtoGeometrySnapshotJson.kt
package expo.modules.kscanlivevtonative

/**
 * Deterministic JSON serialization of a GeometrySnapshot.
 *
 * Hand-rolled rather than org.json on purpose: `org.json` is an Android
 * framework class stubbed out in JVM unit tests, and this serializer has to
 * run in exactly the two places conformance depends on -- the JVM
 * conformance test that writes the native side of the delta table, and the
 * on-device diagnostic that proves the device computes the same numbers.
 * A serializer that only worked in one of them would leave the other's
 * evidence unverifiable.
 *
 * Key order is fixed and floats are emitted at full precision, so two runs
 * of the same input produce byte-identical output and the evidence files
 * diff cleanly.
 */
object GeometrySnapshotJson {

  fun encode(s: GeometrySnapshot, includeMesh: Boolean = false): String {
    val b = StringBuilder(1024)
    b.append('{')
    b.str("fixtureId", s.fixtureId).append(',')
    b.str("bodyFrameId", s.bodyFrameId).append(',')
    b.str("activeAssetId", s.activeAssetId).append(',')
    b.str("assetVersion", s.assetVersion).append(',')
    b.append("\"failure\":").append(s.failure?.let { quote(it) } ?: "null").append(',')
    b.append("\"gatePassed\":").append(s.gatePassed).append(',')
    b.append("\"gateFindings\":[").append(s.gateFindings.joinToString(",") { quote(it) }).append("],")
    b.num("scale", s.scale).append(',')
    b.num("rotationRadians", s.rotationRadians).append(',')
    b.append("\"controlPoints\":{")
    // sorted for determinism -- map iteration order must not leak into evidence
    b.append(s.controlPoints.entries.sortedBy { it.key }.joinToString(",") { (k, v) ->
      "${quote(k)}:[${fmt(v.x)},${fmt(v.y)}]"
    })
    b.append("},")
    b.append("\"bounds\":{")
      .num("minX", s.boundsMin.x).append(',')
      .num("minY", s.boundsMin.y).append(',')
      .num("maxX", s.boundsMax.x).append(',')
      .num("maxY", s.boundsMax.y).append("},")
    b.num("canvasWidth", s.canvasWidth).append(',')
    b.num("canvasHeight", s.canvasHeight).append(',')
    b.append("\"textureWidth\":").append(s.textureWidth).append(',')
    b.append("\"textureHeight\":").append(s.textureHeight).append(',')
    b.append("\"meshWidth\":").append(s.meshWidth).append(',')
    b.append("\"meshHeight\":").append(s.meshHeight).append(',')
    b.append("\"meshVertexCount\":").append(s.meshVertices?.size?.div(2) ?: 0).append(',')
    b.append("\"validationProblems\":[").append(s.validate().joinToString(",") { quote(it) }).append(']')
    if (includeMesh && s.meshVertices != null) {
      b.append(",\"meshVertices\":[").append(s.meshVertices.joinToString(",") { fmt(it) }).append(']')
    }
    b.append('}')
    return b.toString()
  }

  private fun fmt(f: Float): String = if (f.isFinite()) f.toString() else "\"$f\""
  private fun quote(v: String): String = "\"" + v.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
  private fun StringBuilder.str(k: String, v: String) = append(quote(k)).append(':').append(quote(v))
  private fun StringBuilder.num(k: String, v: Float) = append(quote(k)).append(':').append(fmt(v))
}

/** Short one-line form for logcat -- never carries mesh vertices. */
fun describeSnapshot(s: GeometrySnapshot): String = GeometrySnapshotJson.encode(s, includeMesh = false)
